provider/workspace_preset_test.go

```diff
@@ -0,0 +1,127 @@
+				resource := state.Modules[0].Resources["data.coder_workspace_preset.preset_1"]
+				require.NotNil(t, resource)
+				attrs := resource.Primary.Attributes
+				require.Equal(t, attrs["name"], "preset_1")
+				require.Equal(t, attrs["parameters.region"], "us-east1-a")
+				return nil
+			},
+		}},
+	})
+
+	// Given the Name field is not provided
+	resource.Test(t, resource.TestCase{
+		ProviderFactories: coderFactory(),
+		IsUnitTest:        true,
+		Steps: []resource.TestStep{{
+			Config: `
+			data "coder_workspace_preset" "preset_1" {
+				parameters = {
+					"region" = "us-east1-a"
+				}
+			}`,
+			// This is from terraform's validation based on our schema, not based on our validation in ReadContext:
+			ExpectError: regexp.MustCompile("The argument \"name\" is required, but no definition was found"),
+		}},
+	})
+
+	// Given the Name field is empty
+	resource.Test(t, resource.TestCase{
+		ProviderFactories: coderFactory(),
+		IsUnitTest:        true,
+		Steps: []resource.TestStep{{
+			Config: `
+			data "coder_workspace_preset" "preset_1" {
+				name = ""
+				parameters = {
+					"region" = "us-east1-a"
+				}
+			}`,
+			ExpectError: regexp.MustCompile("workspace preset name must be set"),
+		}},
+	})
+
+	// Given the Name field is not a string
+	resource.Test(t, resource.TestCase{
+		ProviderFactories: coderFactory(),
+		IsUnitTest:        true,
+		Steps: []resource.TestStep{{
+			Config: `
+			data "coder_workspace_preset" "preset_1" {
+				name = [1, 2, 3]
+				parameters = {
+					"region" = "us-east1-a"
+				}
+			}`,
+			ExpectError: regexp.MustCompile("Incorrect attribute value type"),
+		}},
+	})
+
+	// Given the Parameters field is not provided
+	resource.Test(t, resource.TestCase{
+		ProviderFactories: coderFactory(),
+		IsUnitTest:        true,
+		Steps: []resource.TestStep{{
+			Config: `
+			data "coder_workspace_preset" "preset_1" {
+				name = "preset_1"
+			}`,
+			ExpectError: regexp.MustCompile("The argument \"parameters\" is required, but no definition was found"),
+		}},
+	})
+
+	// Given the Parameters field is empty
+	resource.Test(t, resource.TestCase{
+		ProviderFactories: coderFactory(),
+		IsUnitTest:        true,
+		Steps: []resource.TestStep{{
+			Config: `
+			data "coder_workspace_preset" "preset_1" {
+				name = "preset_1"
+				parameters = {}
+			}`,
+			ExpectError: regexp.MustCompile("workspace preset must define a value for at least one parameter"),
+		}},
+	})
+
+	// Given the Parameters field is not a map
+	resource.Test(t, resource.TestCase{
+		ProviderFactories: coderFactory(),
+		IsUnitTest:        true,
+		Steps: []resource.TestStep{{
+			Config: `
+			data "coder_workspace_preset" "preset_1" {
+				name = "preset_1"
+				parameters = "not a map"
+			}`,
+			// This is from terraform's validation based on our schema, not based on our validation in ReadContext:
+			ExpectError: regexp.MustCompile("Inappropriate value for attribute \"parameters\": map of string required"),
+		}},
+	})
```

Comment:
Nit: ditto.

Reply:
I've added comments to explain why I'm doing it. TL;DR: While terraform is technically the tool doing the validation, we still need to correctly configure the schema. Regardless of what generates the error message, I'm testing the behaviour.